StratisCore.UI/src/app/login/login.component.ts

```diff
@@ -37,14 +55,16 @@ export class LoginComponent implements OnInit {
       "password": [{value: "", disabled: this.isDecrypting}, Validators.required]
     });
 
-    this.openWalletForm.valueChanges
-      .subscribe(data => this.onValueChanged(data));
+    this.subscriptions.push(this.openWalletForm.valueChanges
```

Comment:
See if we can manage subscriptions a bit better. Under: https://github.com/stratisproject/StratisCore/tree/master/StratisCore.UI/src/app/wallet/tokens/models we have `disposable` implementation. You can see how it is used here: https://github.com/stratisproject/StratisCore/blob/master/StratisCore.UI/src/app/wallet/tokens/components/tokens.component.ts (Mixin + implementing disposable)

Reply:
If you look at the previous code here the subscription was left un-assigned and was not managed at all, this was a potential memory leak. I now add this the subscriptions array and on `ngOnDestroy()` I unsubscribe
```
public ngOnDestroy(): void {
    this.subscriptions.forEach(subscription => subscription.unsubscribe());
  }
```
I have looked at `tokens.component.ts` and read this article 
[https://blog.angularindepth.com/rxjs-avoiding-takeuntil-leaks-fb5182d047ef](https://blog.angularindepth.com/rxjs-avoiding-takeuntil-leaks-fb5182d047ef)

To be honest this code looks very confusing and problematic to me and I am don't really understand the benefit of using this takeUntil operator to manage subscriptions as it's not a common pattern and not obvious to the developer whats happening. You also have to remember to make sure takeUntil is the last operator or the developer will risk a memory leak still.

Angular natively supports management of subscriptions using the template `| async` pipe which I was going to apply to apply more of the code base as a separate exercise to this.

The `tokens.component.ts` also looks far too big IMHO and feels like it is mixing responsibility. I would expect a lot of the work being done in the component should be done in the `TokenService` or `SmartContractService`

Reply:
There are lots of articles about this stuff, but the usual consensus is to use `| async` pipe where possible, if not possible use `takeUntil` and `subscriptions` as last. If you can use `| async` pipe that would be the best.